src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1977,13 +2006,13 @@ fn strip_generics_from_path(path_str: &str) -> Result<String, ResolutionFailure<
     }
 }
 
-fn strip_generics_from_path_segment(segment: &str) -> Result<String, ResolutionFailure<'static>> {
+fn strip_generics_from_path_segment(segment: Vec<char>) -> Result<String, ResolutionFailure<'static>> {
```

Comment:
I don't think this needs to be a `Vec`, you could take `&[char]` or `impl Iterator<Item = char>` to avoid copying. But if you went with the peekable idea I think `&str` would still work.

Reply:
Hmm, I'm not sure if there's a way to avoid the allocation with the `Peekable` approach.

Reply:
How would I slice `path_str` without an allocation?

Reply:
Use `char_indices`, not chars. Then it would look like
```rust
     while let Some((chr, idx)) = path.next() {
        let slice = &path[start..idx - 2];
```
You'd still have to keep track of `start` yourself explicitly, I think it would start at 0 and update to `idx + 1` in every loop iteration.

Reply:
You can slice an iterator?

Reply:
Not `path`, sorry, `path_str`. `let slice = &path_str[start..idx - 2];`

Reply:
Although then I can't use `next_if_eq`... oh well 🤷

Reply:
Hmm, we could run into Unicode issues because `path` will be an `Iterator<char>` but we'll be indexing into a `&[u8]`.

Reply:
`char_indices` guarentees that each index is a valid unicode code point.

Reply:
> Although then I can't use next_if_eq... oh well 🤷

You could use `slice.chars()` again I think. It depends exactly what you code looks like, but you can always turn a `&str` into a `chars()` even if you're starting halfway through the string.

Reply:
Well, it's not that big a deal. I'm just using `matches!`.

Reply:
Is it worth the more complex code to avoid this allocation though?

Reply:
I mean, probably not. If it's difficult to implement I guess it's fine to keep as is, I think you're mostly there though.

Reply:
I think I am there or almost there, but I think I still prefer the way it was before for two reasons:

1. I think the code was simpler and easier to understand
2. I now need to calculate the length of the whole path for the part that processes the last path segment. Before I only needed to compute the length of the current segment being processed

Reply:
The code compiled but I'm now having an issue with how it handles fully-qualified syntax. I think let's stay with the `Peekable` and `segment: String` approach. If perf is an issue, it could be changed in the future since it's an implementation detail.

Reply:
Ok, that seems reasonable.